### PR TITLE
Use the same ulong workaround as in nmod_vec.h

### PR DESCRIPTION
**Pull request by williamstein**
For https://github.com/sagemathinc/JSage, I'm building FLINT for WebAssembly against GMP with long long limbs (WASM has 32-bit pointers but 64-bit arithmetic, so it WASM GMP is much faster when built with long long limbs).  This ends up conflicting with how "ulong" is defined internally in FLINT.   FLINT already has a workaround for this sort of problem in the file nmod_vec.h, but not in qsieve/factor.c.  I'm proposing using the same workaround there.

### CODE REVIEW

**Comment by wbhart** [comment]
Thanks. I'll have to look into this some more. My impression is we use something like this to work around ulong being defined elsewhere, but we don't actually use ulongxx in the rest of the file.

But I'm probably missing that this is solving a different problem which I don't recall the details of.

**Comment by williamstein** [comment]
Thanks.  In this qsieve file you're right that ulong is used twice in the file so this is more subtle.

No hurry, since for my use I'm just applying this as a patch as part of the build process: https://github.com/sagemathinc/JSage/tree/main/packages/flint

**Comment by tthsqe12** [comment]
How is longlong.h working out in wasm?

**Comment by williamstein** [comment]
> How is longlong.h working out in wasm?

It makes a lot of benchmarks run twice as fast, which is nothing to sneeze at...

**Comment by wbhart** [comment]
@williamstein Unfortunately this patch fails in our CI. Unless those system headers that are guarded by this definition actually define ulong, this will not compile, and that would include quite a lot of systems.

In nmod_vec.h we do not use ulongxx anywhere in the file itself. It is only used to *rename* ulong in the guarded system headers so we are then free to redefine ulong (to mp_limb_t) before then using ulong (not ulongxx) in the file.

We tested that the solution in nmod_vec.h does work here (for us). Perhaps you can confirm if it works for you or not.

I believe you are trying to solve a different problem to us. We use this solution to avoid conflict with certain system headers on *some* systems which actually define their own ulong (for entirely different purposes). You appear to be using it to support a system with 32 bit pointers, but 64 bit integers.

I am honestly not sure of the current status of Flint on such systems. In theory any issues that arise on such machines are bugs and should be reported as such. Some people have worked with us over long periods of time to port Flint to such systems, and I have no idea how close we got (or even if we may have finally succeeded). Unfortunately Real Life TM got in the way of them completing the porting work as far as I recall.

So in summary: I'm rejecting the PR as is because it doesn't compile. I'm happy to work with you to support WASM in the way that you want. However, this is not a super priority for us and so we may not be able to respond quickly, as much as I'd like to.

**Comment by wbhart** [comment]
Also at this point I didn't understand why the number of times ulong is used in the file makes a difference here. So there is still something I'm not understanding.

**Comment by williamstein** [comment]
I'll close this, and re-open it if I plan to work on it again.
